### PR TITLE
[Wrangler Ignore option 2] exclude/include logic in wrangler.toml for static asset uploads

### LINKED ISSUE #716: RFC: .wranglerignore
### The Problem
There are static assets directories that may contain lots of content that doesn't need to be uploaded to KV (and likely will trigger KV value size limits). This includes, but is not limited to, things like `node_modules/` and hidden files/directories. Users should be able to specify paths (in the vein of `.gitignore`) that do not get uploaded to KV.

### Option 1: .wranglerignore
We will introduce support for `.wranglerignore` files. These files will be automatically generated with `wrangler init` and `wrangler generate` and placed in the same directory (pwd) as `wrangler.toml`. By default Wrangler will ignore `node_modules` and hidden files (they don't need to be specified in `.wranglerignore`).

These files will be parsed and used exactly like `.gitignore` whenever bucket `upload_files()` is called. The https://docs.rs/ignore/0.4.10/ignore crate will help us a lot here.

### Option 2: include/exclude in wrangler.toml
A more rust-idomatic include/exclude approach is the one used by cargo: https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields-optional

This approach would become part of the `[site]` section of `wrangler.toml`, which would have two new fields, exclude and include. If both are present, wrangler will use include only (same as cargo). The exclude and include fields are used to filter file uploads from whatever directory is specified in the bucket field in wrangler.toml.

The benefits of this approach are that
- Wrangler users will have one fewer file to deal with (no .wranglerignore)
- We can constrain this ignore functionality to just Workers Sites, the only workers application that would use this ignore functionality
- The coupling between Workers sites and include/exclude would neatly be represented in wrangler.toml.

This implementation should be readily doable with https://docs.rs/ignore/0.4.10/ignore/overrides/struct.OverrideBuilder.html.

This approach makes it easier, code-wise, to always ignore node_modules (no messiness about expecting it in the .wranglerignore, we can just always put this logic in the OverrideBuilder).

I advocate for approach number 2. It's fairly easy for me to make implementations of both approaches above, so I can go ahead and the team can play around with the options and tell me which one they like best.

### Proofs of concept
Our two candidates:
1. `.wranglerignore` #720 
1. `include/exclude` in `wrangler.toml` #760

### PR DESCRIPTION
**Pull request by gabbifish**
This PR is the secondary implementation/option to close #716. The first option is #720.

Wrangler will look at `wrangler.toml`'s `site` section for fields to ignore:
```toml
[site]
bucket = "./public"
entry-point = "workers-site" 
include = ["included_dir"] # optional
exclude = ["excluded_dir"] # optional
```
The include and exclude fields will apply to the `bucket` directory. Its behavior above is exactly the same as https://github.com/ashleygwilliams/cargo-generate#include--exclude (this is also the standard for cargo).

Keep in mind that, consistent with other implementations of include/exclude, the include field always overrides the exclude one.

This PR also contains a bunch of unit tests to ensure that our exclude/include logic works as expected.

### CODE REVIEW

File: src/commands/kv/bucket/mod.rs [hunk 1]
```diff
@@ -15,22 +15,24 @@ use std::path::Path;
...
     verbose: bool,
 ) -> Result<(Vec<KeyValuePair>, AssetManifest), failure::Error> {
     let mut upload_vec: Vec<KeyValuePair> = Vec::new();
     let mut asset_manifest: AssetManifest = AssetManifest::new();
 
-    for entry in WalkDir::new(directory)
-        .into_iter()
-        .filter_entry(|e| !is_ignored(e))
-    {
+    let dir_walker = get_dir_iterator(target, directory)?;
```
Comment (EverlastingBugstopper) [inline src/commands/kv/bucket/mod.rs]: nit: is it a dir_walker or a dir_iterator?

File: src/commands/kv/bucket/mod.rs [hunk 2]
```diff
@@ -60,9 +62,12 @@ pub fn directory_keys_values(
...
 
 // Returns only the hashed keys for a directory's files.
-fn directory_keys_only(directory: &Path) -> Result<Vec<String>, failure::Error> {
+fn directory_keys_only(target: &Target, directory: &Path) -> Result<Vec<String>, failure::Error> {
     let mut upload_vec: Vec<String> = Vec::new();
-    for entry in WalkDir::new(directory) {
+
+    let dir_walker = get_dir_iterator(target, directory)?;
+
+    for entry in dir_walker {
```
Comment (EverlastingBugstopper) [inline src/commands/kv/bucket/mod.rs]: what can an `entry` be? mostly just curious

File: src/commands/kv/bucket/mod.rs [hunk 3]
```diff
@@ -79,34 +84,36 @@ fn directory_keys_only(directory: &Path) -> Result<Vec<String>, failure::Error>
...
+        required_override.add(&format!("!{}", ignored))?;
     }
 
-    // Then, ensure files in ignored directories are also ignored.
-    for dir in KNOWN_UNNECESSARY_DIRS {
-        if stem == *dir {
-            // Need to check for full equality here
-            message::info(&format!("ignoring directory {}", dir));
-            return true;
+    if let Some(s) = &target.site {
```
Comment (EverlastingBugstopper) [inline src/commands/kv/bucket/mod.rs]: nit: could you rename `s` to `site`

File: src/commands/kv/bucket/mod.rs (hunk 1 above)
Comment (gabbifish) [reply src/commands/kv/bucket/mod.rs]: i guess they are kind of the same thing--the walker iterates through the directory entries :)

File: src/commands/kv/bucket/mod.rs (hunk 2 above)
Comment (gabbifish) [reply src/commands/kv/bucket/mod.rs]: Entry is any item found within a directory, whether it is a dir or file. it also can include symlinks.

File: src/commands/kv/bucket/mod.rs (hunk 3 above)
Comment (gabbifish) [reply src/commands/kv/bucket/mod.rs]: sure!

**Comment by gabbifish** [comment]
@EverlastingBugstopper responding to questions:
1. There is no output about what's being ignored, given that the user explicitly already states what should be ignored in the `wrangler.toml` file. I can definitely add some info text about `node_modules` being ignored, though.
1. You can't really use `.gitignore` with the paths like `./img`--it's a bit of an antipattern, you usually just provide `img/`. The same applies for this approach with `exclude/include`. This is because these types of ignore patterns can come from a `.ignore` file anywhere in a project, so using relative paths dependent on `.` is unsafe.
1. I can definitely make the error message a bit better and remind folks to use an array if they just provide a string.
1. I'll add info logs for everything ignored. :)

**Comment by EverlastingBugstopper** [comment]
Makes sense - re no. 2, do you think it makes sense to print a `message::warn` when something that is in an `include` or an `exclude` is not recognized? is that possible? like "hey looks like you wanna ignore this but i have no idea what that even is". maybe not necessary but might be helpful. the info logs might be enough for this

**Comment by gabbifish** [comment]
@EverlastingBugstopper I might wait until we see a ticket related to git entries starting with `./` -- I feel like this case is unlikely, but if an issue is filed, I will definitely add additional hand-holding logic!

**Pull request merged**